[BOLT] Match blocks with pseudo probes

shawbyoung opened the pull request:
Match inline trees first between profile and the binary: by GUID,
checksum, parent, and inline site for inlined functions. Map profile
probes to binary probes via matched inline tree nodes. Each binary probe
has an associated binary basic block. If all probes from one profile
basic block map to the same binary basic block, it’s an exact match,
otherwise the block is determined by majority vote and reported as loose
match.

Pseudo probe matching happens between exact hash matching and call/loose
matching.

Introduce ProbeMatchSpec - a mechanism to match probes belonging to
another binary function. For example, given functions foo and bar:
```
void foo() {
  bar();
}
```
profiled binary: bar is not inlined => have top-level function bar
new binary where the profile is applied to: bar is inlined into foo.

Currently, BOLT does 1:1 matching between profile functions and binary
functions based on the name. #100446 will extend this to N:M where
multiple profiles can be matched to one binary function (as in the
example above where binary function foo would use profiles for foo and
bar), and one profile can be matched to multiple binary functions (e.g.
if bar was inlined into multiple functions).

In this diff, ProbeMatchSpecs would only have one BinaryFunctionProfile
(existing name-based matching). 

Test Plan: Added match-blocks-with-pseudo-probes.test

Performance test:
- Setup:
  - Baseline no-BOLT: Clang with pseudo probes, ThinLTO + CSSPGO
  (#79942)
  - BOLT fresh: BOLTed Clang using fresh profile,
  - BOLT stale (hash): BOLTed Clang using stale profile (collected on
    Clang 10K commits back), `-infer-stale-profile` (hash+call block
    matching)
  - BOLT stale (+probe): BOLTed Clang using stale profile,
    `-infer-stale-profile` with `-stale-matching-with-pseudo-probes`
    (hash+call+pseudo probe block matching)
  - 2S Intel SKX Xeon 6138 with 40C/80T and 256GB RAM, using 20C/40T for
    build,
  - BOLT profiles are collected on Clang compiling large preprocessed
    C++ file.
- Benchmark: building Clang (average of 5 runs), see driver in
  aaupov/llvm-devmtg-2022
- Results, wall time, lower is better:
  - Baseline no-BOLT: 429.52 +- 2.61s,
  - BOLT stale (hash): 413.21 +- 2.19s,
  - BOLT stale (+probe): 409.69 +- 1.41s,
  - BOLT fresh: 384.50 +- 1.80s.

shawbyoung committed:
[𝘀𝗽𝗿] initial version

Created using spr 1.3.4

aaupov committed:
[𝘀𝗽𝗿] changes to main this commit is based on

Created using spr 1.3.4

[skip ci]

shawbyoung committed:
Changed assignment of profiles with pseudo probe index

Created using spr 1.3.4

shawbyoung committed:
Edit test and assert

Created using spr 1.3.4

aaupov commented on bolt/lib/Profile/StaleProfileMatching.cpp:
```diff
@@ -266,6 +287,47 @@ class StaleMatcher {
+      return nullptr;
+
+    uint64_t Index = BlockPseudoProbes[0]->Index;
+    assert(Index <= Blocks.size() && "Invalid pseudo probe index");
```
Invalid assertion

aaupov commented on bolt/lib/Profile/StaleProfileMatching.cpp:
```diff
@@ -478,11 +544,44 @@ matchWeightsByHashes(BinaryContext &BC,
   }
+
+  uint64_t BFPseudoProbeDescHash = 0;
+  if (BF.hasPseudoProbe()) {
```
This is a misleading interface that we should drop/move to BC. 
Please use `BF.getGUID() != 0` as a check for whether BF has pseudo probes.

aaupov commented on bolt/lib/Profile/StaleProfileMatching.cpp:
```diff
@@ -266,6 +287,47 @@ class StaleMatcher {
+
+    auto It = IndexToBinaryPseudoProbes.find(Index);
+    assert(It != IndexToBinaryPseudoProbes.end() &&
+           "All blocks should have a pseudo probe");
```
This assert should become a check as it's possible to have blocks without probes

shawbyoung committed:
Fixed failing asserts, pruned prospective pseudo probes for matching

Created using spr 1.3.4

aaupov commented on bolt/lib/Profile/StaleProfileMatching.cpp:
```diff
@@ -555,6 +574,10 @@ size_t matchWeightsByHashes(
                            ProbeMap.lower_bound(FuncAddr + BlockRange.second));
       for (const auto &[_, Probes] : BlockProbes) {
         for (const MCDecodedPseudoProbe &Probe : Probes) {
+          if (Probe.getInlineTreeNode()->hasInlineSite())
```
What do we prune with this check? Don't we discard valid probes belonging to the current function?

shawbyoung replied:
This pruning resulted in a tangible increase in 1:1 mappings btw profile and binary pseudo probes - the PseudoProbeDecoder::ProbeMap interface not only contains the probe attached to some block b, but contains the block probes attached to block b inlined in other functions.

aaupov commented on bolt/lib/Profile/StaleProfileMatching.cpp:
```diff
@@ -306,26 +310,41 @@ class StaleMatcher {
-    if (BlockPseudoProbes.size() == 0 || BlockPseudoProbes.size() > 1)
+    if (BlockPseudoProbes.size() == 0) {
+      if (opts::Verbosity >= 2)
+        errs() << "BOLT-WARNING: no pseudo probes in profile block\n";
```
Bump verbosity for this logging to >=3.
Add aggregated counters – those could be printed for BF at v>=2.
BC-level aggregated counters can be printed at v>=1.

shawbyoung committed:
Added logging for pseudo probe block matching

Created using spr 1.3.4

shawbyoung committed:
Changed pseudo probe matching failure logging to v=3

Created using spr 1.3.4

shawbyoung committed:
More loggin

Created using spr 1.3.4

shawbyoung commented:
cc @WenleiHe @wlei-llvm

shawbyoung committed:
Logging blocks matched with opcodes

Created using spr 1.3.4

shawbyoung committed:
Updated test

Created using spr 1.3.4

shawbyoung committed:
Name changes in prep for inlined block pseudo probe block matching

Created using spr 1.3.4

shawbyoung committed:
Rm unnecessary Blocks vec in StaleMatcher

Created using spr 1.3.4

shawbyoung committed:
Improved matched block counting

Created using spr 1.3.4

shawbyoung committed:
Removed comment from test

Created using spr 1.3.4

shawbyoung committed:
Added comments and check for null YamlBFGUID in StaleMatcher before PseudoProbe matching

Created using spr 1.3.4

aaupov commented on bolt/lib/Profile/StaleProfileMatching.cpp:
```diff
@@ -206,13 +213,30 @@ class StaleMatcher {
         CallHashToBlocks[CallHashes[I]].push_back(
             std::make_pair(Hashes[I], Block));
     }
+    this->IndexToBBPseudoProbes = IndexToBBPseudoProbes;
+    this->BBPseudoProbeToBlock = BBPseudoProbeToBlock;
+    this->YamlBFGUID = YamlBFGUID;
```
As we only perform pass-through assignment with these members, would it make sense to set them in the constructor or simply expose as public?

shawbyoung replied:
Made some changes to the stale matcher class in response to this - added a constructor and some setters. Might want to follow up in a separate PR and make similar NFC changes to OpHashToBlocks and CallHashToBlocks. Let me know what you think.

aaupov replied:
Yes, that looks good.

aaupov commented on bolt/lib/Profile/StaleProfileMatching.cpp:
```diff
@@ -206,13 +213,30 @@ class StaleMatcher {
+      return BestBlock;
+    }
+    BestBlock = matchWithCalls(BlendedHash, CallHash);
+    if (BestBlock) {
+      return BestBlock;
+    }
```
I see the intent here but let's stick to the code style and omit braces

aaupov commented on bolt/lib/Profile/StaleProfileMatching.cpp:
```diff
@@ -447,18 +550,22 @@ createFlowFunction(const BinaryFunction::BasicBlockOrderType &BlockOrder) {
+      IndexToBBPseudoProbes;
+  std::unordered_map<const MCDecodedPseudoProbe *, FlowBlock *>
+      BBPseudoProbeToBlock;
+  const MCPseudoProbeDecoder *PseudoProbeDecoder = BC.getPseudoProbeDecoder();
```
Please put this under an option https://github.com/llvm/llvm-project/pull/100299

aaupov reviewed (commented):
Looking good with a couple of nits.

shawbyoung committed:
Omitting braces in one line if

Created using spr 1.3.4

aaupov committed:
[𝘀𝗽𝗿] changes introduced through rebase

Created using spr 1.3.4

[skip ci]

shawbyoung committed:
Pseudo probe matching now is triggered by flag

Created using spr 1.3.4

shawbyoung committed:
Omit unnecessary braces

Created using spr 1.3.4

shawbyoung committed:
Change initialization of index -> probe and probe -> block mappings

Created using spr 1.3.4

shawbyoung committed:
Formatting

Created using spr 1.3.4

shawbyoung committed:
Comments

Created using spr 1.3.4

aaupov reviewed (approved):
LG but let's give some time for other reviewers

ayermolo commented on bolt/lib/Profile/StaleProfileMatching.cpp:
```diff
@@ -266,6 +313,65 @@ class StaleMatcher {
+      return nullptr;
+
+    if (opts::Verbosity >= 3)
+      outs() << "BOLT-INFO: attempting to match block with pseudo probes\n";
```
Aren't we now calling outs()/errs() from BinaryContext?

aaupov replied:
Not throughout Profile component (yet).

dcci commented on bolt/lib/Profile/StaleProfileMatching.cpp:
```diff
@@ -223,10 +249,31 @@ class StaleMatcher {
   using HashBlockPairType = std::pair<BlendedBlockHash, FlowBlock *>;
   std::unordered_map<uint16_t, std::vector<HashBlockPairType>> OpHashToBlocks;
   std::unordered_map<uint64_t, std::vector<HashBlockPairType>> CallHashToBlocks;
+  std::unordered_map<uint64_t, std::vector<const MCDecodedPseudoProbe *>>
```
usual Q, how come we're using `std::unordered` rather than an LLVM ADT?

shawbyoung replied:
I wanted to keep the mappings used within StaleMatcher consistent, but since I'm initializing pseudo probe mappings and iterating on the StaleMatcher differently anyways, I'll change these to LLVM ADT.

WenleiHe replied:
@dcci DenseMap or LLVM ADT isn't generally better. we have a ton of undered_map throughout the code base, and we have benchmarked some cases carefully between the two, and DenseMap sometimes uses a lot more memory, as a result many uses of undered_map were intentional. 

If memory is of a concern (probe decoding and holding probe forest generally isn't cheap), I'm not sure DenseMap will help us.

dcci replied:
It's not better, but there's a use case for both of them. My question is mostly about why we use one instead of the other.

spupyrev replied:
@dcci @shawbyoung this due to a limitation of ADT maps, e.g., the following code crashes:
```  
  DenseMap<uint16_t, uint64_t> TestMap;
  TestMap[uint16_t(-1)] = 123;
```
If not that, I agree that ADT data structures would be preferred

dcci commented on bolt/lib/Profile/StaleProfileMatching.cpp:
```diff
@@ -266,6 +313,65 @@ class StaleMatcher {
+    // Searches for the pseudo probe attached to the matched function's block,
+    // ignoring pseudo probes attached to function calls and inlined functions'
+    // blocks.
+    std::vector<const yaml::bolt::PseudoProbeInfo *> BlockPseudoProbes;
```
SmallVector?

dcci commented on bolt/lib/Profile/StaleProfileMatching.cpp:
```diff
@@ -266,6 +313,65 @@ class StaleMatcher {
+        errs() << "BOLT-WARNING: no pseudo probes in profile block\n";
+      return nullptr;
+    }
+    if (BlockPseudoProbes.size() > 1) {
```
you can probably hoist the computation of size and compute it only once.

dcci commented on bolt/lib/Profile/StaleProfileMatching.cpp:
```diff
@@ -266,6 +313,65 @@ class StaleMatcher {
+      return nullptr;
+    }
+    if (It->second.size() > 1) {
+      if (opts::Verbosity >= 3)
```
there's some repetition here 
```
if (verbosity >= 3) { ... }
```

I wonder if you can abstract this in a common helper that would help you simplifying the code.

dcci reviewed (commented):
some comments.

WenleiHe commented on bolt/lib/Profile/StaleProfileMatching.cpp:
```diff
@@ -478,10 +605,31 @@ matchWeightsByHashes(BinaryContext &BC,
+            continue;
+          if (Probe.getType() != static_cast<uint8_t>(PseudoProbeType::Block))
+            continue;
+          Matcher.mapIndexToProbe(Probe.getIndex(), &Probe);
+          Matcher.mapProbeToBB(&Probe, Blocks[I]);
```
Where do we guard all of this by CFG hash check? I was looking for something like `BFPseudoProbeDescHash == YamlBF.PseudoProbeDescHash`.

I can see `YamlBFGUID` is being checked in `matchWithPseudoProbes`, but 1) this isn't obvious from readability POV, 2) we don't need to populate the mapping if we are not gonna trust probe at all (`BFPseudoProbeDescHash == YamlBF.PseudoProbeDescHash`).

shawbyoung replied:
Addressed in addition to inlined block pseudo probe matching.

WenleiHe commented on bolt/lib/Profile/StaleProfileMatching.cpp:
```diff
@@ -478,10 +605,31 @@ matchWeightsByHashes(BinaryContext &BC,
+                           ProbeMap.lower_bound(FuncAddr + BlockRange.second));
+      for (const auto &[_, Probes] : BlockProbes) {
+        for (const MCDecodedPseudoProbe &Probe : Probes) {
+          if (Probe.getInlineTreeNode()->hasInlineSite())
```
If we think probe is a reliable anchor for block level profile-binary matching, we should find ways to leverage inlinee probes for matching inlinee blocks too. Operating with only top level probes can be very limiting.

shawbyoung replied:
Just added inlined block pseudo probe matching.

WenleiHe commented on bolt/lib/Profile/StaleProfileMatching.cpp:
```diff
@@ -208,11 +212,33 @@ class StaleMatcher {
+    BestBlock = matchWithCalls(BlendedHash, CallHash);
+    if (BestBlock)
+      return BestBlock;
+    BestBlock = matchWithPseudoProbes(BlendedHash, PseudoProbes);
```
I'm a bit unsure about the strategy here. This is assuming matching of each individual block/node is an isolated problem. But the reality is, it's not, instead we're matching two sets of nodes (or rather two graphs) holistically.

I.e. if we function's CFG hash matches that from probe desc, we should be able to have high confidence exact mapping for all blocks based on probe, and shouldn't need to treat probe based matching as a fallback?

aaupov replied:
> I'm a bit unsure about the strategy here. This is assuming matching of each individual block/node is an isolated problem. But the reality is, it's not, instead we're matching two sets of nodes (or rather two graphs) holistically.

Agree but we're trying to lay the foundation with only block matching belonging to the current function. Your comment makes me rethink probe filtering based on GUID - cc @shawbyoung.

> I.e. if we function's CFG hash matches that from probe desc, we should be able to have high confidence exact mapping for all blocks based on probe, and shouldn't need to treat probe based matching as a fallback?

Exact hash provides a stronger signal than probes and is available for all binary basic blocks, so it makes sense to put it to the top. I'm on the fence about the relative strengths of call hash and probes: call hash is supposedly cheaper to construct but might not be available for all blocks. Stats from both should help us make the right call.

WenleiHe replied:
> Exact hash provides a stronger signal than probes and is available for all binary basic blocks, so it makes sense to put it to the top.

Not necessarily. The tricky bit is for better overall graph matching, strict exact match for a few individual block does't guarantee global best match. We need to capture information that is relevant to CFG change, capture everything in a hash (strict) could also cause matching to be skewed by certain changes that can be considered noise for our purpose. 

Btw, I'm not asking for an immediate change here, but at least something we need to think about, and experiment. In an alternative design, we can have several matching strategy, controlled by a switch, and we can then experiment with them using the same profile. The selection of strategy can be global instead of on per-block basis, at least for a subset of the strategies.

WenleiHe commented on bolt/lib/Profile/StaleProfileMatching.cpp:
```diff
@@ -45,6 +45,8 @@ namespace opts {
 extern cl::opt<bool> TimeRewrite;
 extern cl::OptionCategory BoltOptCategory;
+extern cl::opt<unsigned> Verbosity;
+extern cl::opt<bool> ProfileUsePseudoProbes;
```
IIUC `ProfileUsePseudoProbes` controls whether we extract probe from binary and write it out in YAML profile today. It might be confusing to repurpose this same flag to control stale profile matching strategy. Something specific for profile matching strategy might be better.

aaupov replied:
No, I intended `ProfileUsePseudoProbes` to cover both generation and use of pseudo probes. The option means we parse pseudo probes for use in Profile, and that covers both YAMLProfileReader and Writer.

WenleiHe replied:
Ok, I think we need a dedicated flag to control matching strategy, otherwise it can be confusing. 

Generating profile with probe and using profile for stale profile matching are two different things. Like profile contains probe doesn't mean stale matching has to use probe.

aaupov replied:
> Ok, I think we need a dedicated flag to control matching strategy, otherwise it can be confusing. Generating profile with probe and using profile for stale profile matching are two different things.

Agree they're different but both require parsing pseudo probes, which is controlled by the flag. `ProfileUseDFS` is another flag which controls both profile generation and consumption.

> Like profile contains probe doesn't mean stale matching has to use probe.

If the profile contains probe information, but ProfileUsePseudoProbes is not set, we will not use the information for matching.

WenleiHe replied:
> If the profile contains probe information, but ProfileUsePseudoProbes is not set, we will not use the information for matching.

But this is the confusion... The profile (not stale matching) is clearly using probe as probe exists in profile, why you want users to use such a flag (set to false) to control stale matching strategy. Wouldn't it be better to have a dedicated flag whose name actually reflects its purpose in this case?

aaupov replied:
> The profile (not stale matching) is clearly using probe as probe exists in profile

But in a similar way, the profile can contain BB hashes, but they won't be used unless `-infer-stale-profile` is set. (We produce them unconditionally so there's no corresponding generation flag).

There's no use of probes outside stale matching.

The confusion argument for me is flipped – if there were separate flags for generation and consumption for probes. It's as if we had `ProfileEmitDFSIndices` and `ProfileAssumeDFSIndices` instead of `ProfileUseDFS` which covers both.

WenleiHe replied:
> There's no use of probes outside stale matching. 

True, but you do have the options to let probe live in yaml profile and then not use it (even just for testing). you do have a state this is NOT "having probe and use it for profile matching". 

Wouldn't a switch like `enable-probe-based-profile-matching` reflect the intention much better? We may even need something like `profile-matching-strategy={probe,hash,call}` in the future. 

> It's as if we had ProfileEmitDFSIndices and ProfileAssumeDFSIndices instead of ProfileUseDFS which covers both.

This is different, DFS order is a property of the profile. Whether you use probe to do matching is not.

shawbyoung replied:
Added a flag for pseudo probe block matching.

shawbyoung committed:
Changed std ADTs to LLVM

Created using spr 1.3.4

shawbyoung committed:
In matchWithPseudoProbe, hoist BlocksPseudoProbes.size(), added logging helpr

Created using spr 1.3.4

shawbyoung committed:
A more beautiful helper function for matchWithPseudoProbes

Created using spr 1.3.4

shawbyoung committed:
Added inlined block pseudo probe matching

Created using spr 1.3.4

shawbyoung committed:
Added flag to trigger pseudo probe block matching

Created using spr 1.3.4

wlei-llvm commented on bolt/lib/Profile/StaleProfileMatching.cpp:
```diff
@@ -116,6 +118,11 @@ cl::opt<unsigned> StaleMatchingCostJumpUnknownFTInc(
 
+cl::opt<bool> StaleMatchingWithBlockPseudoProbes(
+    "stale-matching-with-block-pseudo-probes",
+    cl::desc("Turns on stale matching with block pseudo probes."), cl::init(3),
```
`cl::init(3)` should it be bool value?

wlei-llvm commented on bolt/lib/Profile/StaleProfileMatching.cpp:
```diff
@@ -266,6 +325,123 @@ class StaleMatcher {
   }
+
+  /// A helper function for logging.
+  static bool LogErrIfExpr(bool Expr, std::string Message) {
```
Nit: `std::string` can this be `StringRef`?

wlei-llvm commented on bolt/lib/Profile/StaleProfileMatching.cpp:
```diff
@@ -478,10 +675,34 @@ matchWeightsByHashes(BinaryContext &BC,
+          if (Probe.getType() != static_cast<uint8_t>(PseudoProbeType::Block))
+            continue;
+          if (Probe.getInlineTreeNode()->hasInlineSite())
+            Matcher.mapGUIDAndIndexToProbe(Probe.getGuid(), Probe.getIndex(),
+                                           &Probe);
+          else
+            Matcher.mapIndexToProbe(Probe.getIndex(), &Probe);
```
Wondering why we need to use two containers? iiuc, all the probes contain the `GUID`, the top-level probes should contain the function GUID.

shawbyoung replied:
I separated the block probes into two containers to allows for O(1) look up for either, with one less layer of indirection for top level probes. In one large binary, I saw top level probes accounted for ~75% of matched probes, so I thought this was reasonable.

wlei-llvm replied:
It's surprised to me that there are 75% of top level probes, which means compiler inlining introduces only 25% probes/blocks, but in fact, PGO inlining is aggressive which can inline 12X more size(https://github.com/llvm/llvm-project/blob/main/llvm/lib/Transforms/IPO/SampleProfile.cpp#L207) than the original function(the top-level probes).

Even to use two-level look up, that's still constant time complexity, it should not be an issue? I saw large duplicated codes in `matchWithInlinedBlockPseudoProbes` and `matchWithNonInlinedBlockPseudoProbes`, using one container and unify them can be good for the future maintenance.   And also wondering if the map is for other use or just look up? If no other use, how about use combine hash(std::pair<...>(GUID, Index) as a key, then we don't need to use the 2-D DenseMap.

shawbyoung replied:
Hm perhaps the binary I looked at wasn't representative. Thanks for your comment. I think I was thinking in too granular in terms of map optimization, I'll change this.

wlei-llvm commented on bolt/test/X86/match-blocks-with-pseudo-probes.test:
```diff
@@ -0,0 +1,62 @@
+  pushq   %rbp
+  movq    %rsp, %rbp
+  movl    $0, -4(%rbp)
+  .pseudoprobe    15822663052811949562 1 0 0 main
```
Consider to add inlining case? (I guess one big reason we want to use pseudo-probe is to deal with inlining)

shawbyoung committed:
Added flag for pseudo probe block matching

Created using spr 1.3.4

shawbyoung committed:
Set flag init val, changed std::string to StringRef

Created using spr 1.3.4

wlei-llvm commented on bolt/lib/Profile/StaleProfileMatching.cpp:
```diff
@@ -266,6 +325,122 @@ class StaleMatcher {
+                     "BOLT-WARNING: no pseudo probes in profile block\n"))
+      return nullptr;
+    if (LogErrIfExpr(
+            NInlinedBlockPseudoProbes > 1,
```
it should be common that one basic block contains >1 pseudo probes, e.g. when inlining happens, the inlinee's entry pseudo probe(ID:1) is inlined to replace the inline callsite, whose basic block contains the inliner's pseudo probe.

aaupov committed:
[MC][NFC] Drop unused MCDecodedPseudoProbeInlineTree::ChildrenToProcess (#100576)

The usage was removed in 3f97016857b0305294f3a55ea220884fb50ce033.

Results in a slight peak RSS reduction in
`perf2bolt --profile-use-pseudo-probes` from 17.24 to 16.85 GiB.

wlei-llvm commented on bolt/lib/Profile/StaleProfileMatching.cpp:
```diff
@@ -266,6 +325,122 @@ class StaleMatcher {
+                     "BOLT-WARNING: no pseudo probes in profile block\n"))
+      return nullptr;
+    if (LogErrIfExpr(
+            NBlockPseudoProbes > 1,
```
Same here

aaupov committed:
buildAddress2ProbeMap timers

aaupov committed:
[BOLT][NFC] Add timers for MetadataManager invocations



Test Plan: added bolt/test/timers.c

Reviewers: ayermolo, maksfb, rafaelauler, dcci

Reviewed By: dcci

Pull Request: https://github.com/llvm/llvm-project/pull/101267

aaupov committed:
[profgen][NFC] Pass parameter as const_ref

Pass `ProbeNode` parameter of `trackInlineesOptimizedAway` as const
reference.

Reviewers: wlei-llvm, WenleiHe

Reviewed By: WenleiHe

Pull Request: https://github.com/llvm/llvm-project/pull/102787

aaupov committed:
[MC][profgen][NFC] Expand auto for MCDecodedPseudoProbe

Expand autos in select places in preparation to #102789.

Reviewers: dcci, maksfb, WenleiHe, rafaelauler, ayermolo, wlei-llvm

Reviewed By: WenleiHe, wlei-llvm

Pull Request: https://github.com/llvm/llvm-project/pull/102788

aaupov committed:
[MC][NFC] Count pseudo probes and function records

Pre-parse pseudo probes section counting the number of probes and
function records. These numbers are used in follow-up diff to
pre-allocate vectors for decoded probes and inline tree nodes.

Additional benefit is avoiding error handling during parsing.

This pre-parsing is fast: for a 404MiB .pseudo_probe section with
43373881 probes and 25228770 function records, it only takes 0.68±0.01s.
The total time of buildAddress2ProbeMap is 21s.

Reviewers: dcci, maksfb, rafaelauler, wlei-llvm, ayermolo

Reviewed By: wlei-llvm

Pull Request: https://github.com/llvm/llvm-project/pull/102774

aaupov committed:
[MC][NFC] Statically allocate storage for decoded pseudo probes and function records

Use #102774 to allocate storage for decoded probes (`PseudoProbeVec`)
and function records (`InlineTreeVec`).

Leverage that to also shrink sizes of `MCDecodedPseudoProbe`:
- Drop Guid since it's accessible via `InlineTree`.

`MCDecodedPseudoProbeInlineTree`:
- Keep track of probes and inlinees using `ArrayRef`s now that probes
  and function records belonging to the same function are allocated
  contiguously.

This reduces peak RSS from 13.7 GiB to 9.7 GiB and pseudo probe parsing
time (as part of perf2bolt) from 15.3s to 9.6s for a large binary with
400MiB .pseudo_probe section containing 43M probes and 25M function
records.

Depends on:
#102774
#102787
#102788

Reviewers: maksfb, rafaelauler, dcci, ayermolo, wlei-llvm

Reviewed By: wlei-llvm

Pull Request: https://github.com/llvm/llvm-project/pull/102789

aaupov committed:
[MC][NFC] Reduce Address2ProbesMap size

Replace the map from addresses to list of probes with a flat vector
containing probe references sorted by their addresses.

Reduces pseudo probe parsing time from 9.56s to 8.59s and peak RSS from
9.66 GiB to 9.08 GiB as part of perf2bolt processing a large binary.

Test Plan:
```
bin/llvm-lit -sv test/tools/llvm-profgen
```

Reviewers: maksfb, rafaelauler, dcci, ayermolo, wlei-llvm

Reviewed By: wlei-llvm

Pull Request: https://github.com/llvm/llvm-project/pull/102904

aaupov committed:
[MC][NFC] Use vector for GUIDProbeFunctionMap

Replace unordered_map with a vector. Pre-parse the section to statically
allocate storage. Use BumpPtrAllocator for FuncName strings, keep
StringRef in FuncDesc.

Reduces peak RSS of pseudo probe parsing from 9.08 GiB to 8.89 GiB as
part of perf2bolt with a large binary.

Test Plan:
```
bin/llvm-lit -sv test/tools/llvm-profgen
```

Reviewers: wlei-llvm, rafaelauler, dcci, maksfb, ayermolo

Reviewed By: wlei-llvm

Pull Request: https://github.com/llvm/llvm-project/pull/102905

aaupov committed:
[BOLT] Only parse probes for profiled functions in profile-write-pseudo-probes mode

aaupov committed:
[BOLT][NFCI] Strip suffix in getLTOCommonName

Also provide a mechanism to override the list of suffixes to consider.
Override LTOSuffixes for getGUID in pseudo probe parsing.

Test Plan: 

Reviewers:
Subscribers:

Tasks:

Tags:

Differential Revision: https://phabricator.intern.facebook.com/D61857819

Pull Request: https://github.com/llvm/llvm-project/pull/106243

aaupov committed:
[BOLT][NFC] Rename profile-use-pseudo-probes

aaupov committed:
Reworked block probe matching

Use new probe ifaces
Get all function probes at once
Drop ProfileUsePseudoProbes
Unify matchWithBlockPseudoProbes
Distinguish exact and loose probe match

aaupov committed:
[BOLT] Add pseudo probe inline tree to YAML profile

To be used for pseudo probe function matching (#100446).

Test Plan: updated pseudoprobe-decoding-inline.test

Pull Request: https://github.com/llvm/llvm-project/pull/107137

aaupov committed:
[𝘀𝗽𝗿] changes introduced through rebase

Created using spr 1.3.4

[skip ci]

aaupov committed:
rebase

Created using spr 1.3.4

aaupov committed:
drop logIf

Created using spr 1.3.4

aaupov committed:
[𝘀𝗽𝗿] changes introduced through rebase

Created using spr 1.3.4

[skip ci]

aaupov committed:
Use new profile probe encoding

Created using spr 1.3.4

aaupov committed:
[𝘀𝗽𝗿] changes introduced through rebase

Created using spr 1.3.4

[skip ci]

aaupov committed:
Test fix

Created using spr 1.3.4

aaupov committed:
[𝘀𝗽𝗿] changes introduced through rebase

Created using spr 1.3.4

[skip ci]

aaupov committed:
Memoize top-level GUID->InlineTree mapping, cuts inference time by ~30%

Created using spr 1.3.4

aaupov committed:
[𝘀𝗽𝗿] changes introduced through rebase

Created using spr 1.3.4

[skip ci]

aaupov committed:
clang-format

Created using spr 1.3.4

aaupov committed:
[𝘀𝗽𝗿] changes introduced through rebase

Created using spr 1.3.4

[skip ci]

aaupov committed:
Update with #107137 changes

Created using spr 1.3.4

aaupov committed:
[𝘀𝗽𝗿] changes introduced through rebase

Created using spr 1.3.4

[skip ci]

aaupov committed:
Move matchInlineTrees into InlineTreeNodeMapTy

Created using spr 1.3.4

wlei-llvm commented on bolt/lib/Profile/StaleProfileMatching.cpp:
```diff
@@ -266,6 +305,69 @@ class StaleMatcher {
+        return nullptr;
+      const MCDecodedPseudoProbe *BinaryProbe = nullptr;
+      for (const MCDecodedPseudoProbe &Probe : Node->getProbes()) {
+        if (Probe.getIndex() != ProbeId)
```
nit:  just for the readability, is it possible to add the `Binary` or `Profile` prefix to the variable names, especially when they are used closely. Same to other places like `Node/Tree/Probe/...` but only when you think it's not too verbose:)

wlei-llvm commented on bolt/include/bolt/Profile/YAMLProfileReader.h:
```diff
@@ -96,6 +101,42 @@ class YAMLProfileReader : public ProfileReaderBase {
+    // \p GetRootCallback is invoked for matching up the first YAML inline tree
+    // node and has the following signature:
+    // const MCDecodedPseudoProbeInlineTree *GetRootCallback(uint64_t RootGUID)
+    void matchInlineTrees(
```
Maybe rename to `readAndMatchInlineTrees` or `buildAndMatchingInlinTrees`? I see a big part of the code is to read and build the trees from yaml. Or maybe split it into two functions?

wlei-llvm commented on bolt/include/bolt/Profile/YAMLProfileReader.h:
```diff
@@ -96,6 +101,42 @@ class YAMLProfileReader : public ProfileReaderBase {
 
+  // A class for matching inline tree nodes between profile and binary.
+  class InlineTreeNodeMapTy {
+    DenseMap<uint32_t, const MCDecodedPseudoProbeInlineTree *> Map;
```
add comments to explain the key and the value of the map.

wlei-llvm commented on bolt/include/bolt/Profile/YAMLProfileReader.h:
```diff
@@ -96,6 +101,42 @@ class YAMLProfileReader : public ProfileReaderBase {
+  class InlineTreeNodeMapTy {
+    DenseMap<uint32_t, const MCDecodedPseudoProbeInlineTree *> Map;
+
+    void mapInlineTreeNode(uint32_t ProfileNode,
```
nit: `ProfileNode` --> `ProfileNodeIdx`

wlei-llvm commented on bolt/lib/Profile/StaleProfileMatching.cpp:
```diff
@@ -266,6 +305,69 @@ class StaleMatcher {
     return BestBlock;
   }
+
+  /// Matches a profile block with an binary block based on pseudo probes.
```
nit: `an` -> `a`

aaupov commented on bolt/lib/Profile/YAMLProfileReader.cpp:
```diff
@@ -587,6 +590,51 @@ size_t YAMLProfileReader::matchWithCallGraph(BinaryContext &BC) {
+      Cur = GetRootCallback(GUID);
+    } else if (const MCDecodedPseudoProbeInlineTree *Parent =
+                   getInlineTreeNode(ParentId)) {
+      for (const MCDecodedPseudoProbeInlineTree &Child :
```
Sure. But actually can we rely on the fact that children are ordered by their call site probe id? That would make binary search possible.

wlei-llvm replied:
I see, I just remembered it has been changed to a vector structure not a map, then that needs the pre-sorting things. OK, I'm good with the current version assuming the perf here is fine.

aaupov commented on bolt/lib/Profile/YAMLProfileReader.cpp:
```diff
@@ -587,6 +590,51 @@ size_t YAMLProfileReader::matchWithCallGraph(BinaryContext &BC) {
+      PrevGUIDIdx = GUIDIdx;
+    else
+      GUIDIdx = PrevGUIDIdx;
+    assert(GUIDIdx < YamlPD.GUID.size());
```
I'll drop these assertions as they are done by operator[] anyway.

aaupov commented on bolt/lib/Profile/YAMLProfileReader.cpp:
```diff
@@ -587,6 +590,51 @@ size_t YAMLProfileReader::matchWithCallGraph(BinaryContext &BC) {
+        }
+      }
+    }
+    if (Cur && Decoder.getFuncDescForGUID(GUID)->FuncHash == Hash)
```
`Hash` comes from YAML, we compare it against binary checksum.

wlei-llvm replied:
I see, please add comments to explain this is the case for the stale profile, we don't match the probes whose function checksum is mismatched.

aaupov commented on bolt/lib/Profile/YAMLProfileReader.cpp:
```diff
@@ -722,6 +770,15 @@ Error YAMLProfileReader::readProfile(BinaryContext &BC) {
     }
   }
 
+  if (opts::StaleMatchingWithPseudoProbes) {
+    const MCPseudoProbeDecoder *Decoder = BC.getPseudoProbeDecoder();
+    assert(Decoder &&
+           "If pseudo probes are in use, pseudo probe decoder should exist");
+    for (const MCDecodedPseudoProbeInlineTree &TopLev :
+         Decoder->getDummyInlineRoot().getChildren())
+      TopLevelGUIDToInlineTree[TopLev.Guid] = &TopLev;
+  }
+
```
We want to set TopLevelGUID... once though, that's why I decided to keep it in top-level location. matchWeightsByHashes would be invoked for every profile basic block.

aaupov committed:
[𝘀𝗽𝗿] changes introduced through rebase

Created using spr 1.3.4

[skip ci]

aaupov committed:
Add test with inline trees, address comments

Created using spr 1.3.4

aaupov committed:
Allow null block participate in majority vote, improves run-time performance

Created using spr 1.3.4

aaupov commented:
Ping @wlei-llvm

wlei-llvm commented:
> Ping @wlei-llvm

Sorry for the delay. The new version addressed my last comment (with just minor nits). However, I didn't fully follow the new features related to `ProbeMatchSpecs` stuffs. Could you add more descriptions to the diff summary? Or if it’s not a lot of work, could we split it into two patches? We could commit the first part, and I will review the second part separately.

wlei-llvm commented on bolt/lib/Profile/StaleProfileMatching.cpp:
```diff
@@ -447,12 +561,12 @@ createFlowFunction(const BinaryFunction::BasicBlockOrderType &BlockOrder) {
-                     const yaml::bolt::BinaryFunctionProfile &YamlBF,
-                     FlowFunction &Func, HashFunction HashFunction,
-                     YAMLProfileReader::ProfileLookupMap &IdToYamlBF) {
+size_t matchWeightsByHashes(
```
Should we rename this function? Now it's not only "ByHashes", if so, also update the comments above.

wlei-llvm commented on bolt/lib/Profile/StaleProfileMatching.cpp:
```diff
@@ -482,11 +596,68 @@ matchWeightsByHashes(BinaryContext &BC,
+      if (!MatchedBlock)
+        continue;
+      // Don't override earlier matches
+      if (MatchedFlowBlocks.contains(MatchedBlock))
+        continue;
+      MatchedFlowBlocks.try_emplace(MatchedBlock, Method);
+      MatchedBlocks[&YamlBP][YamlBB.Index] = {MatchedBlock, &YamlBB};
```
Those lines are the same for the three matchings, probably can be factored out into a helper function.

wlei-llvm commented:
> > Ping @wlei-llvm
> 
> Sorry for the delay. The new version addressed my last comment (with just minor nits). However, I didn't fully follow the new features related to `ProbeMatchSpecs` stuffs. Could you add more descriptions to the diff summary? Or if it’s not a lot of work, could we split it into two patches? We could commit the first part, and I will review the second part separately.

NVM, I think now I get what `ProbeMatchSpecs` does, it's a vector because a function can have multiple sections(function split)

aaupov commented:
> > > Ping @wlei-llvm
> > 
> > 
> > Sorry for the delay. The new version addressed my last comment (with just minor nits). However, I didn't fully follow the new features related to `ProbeMatchSpecs` stuffs. Could you add more descriptions to the diff summary? Or if it’s not a lot of work, could we split it into two patches? We could commit the first part, and I will review the second part separately.
> 
> NVM, I think now I get what `ProbeMatchSpecs` does, it's a vector because a function can have multiple sections(function split)

Thank you for reviewing and sorry for the delay from my end, was busy with profile quality work.

ProbeMatchSpecs is a mechanism to match probes belonging to another binary function. I'm going to utilize it in probe-based function matching (#100446). For example:

source function:
```
void foo() {
  bar();
}
```

profiled binary: bar is not inlined => have top-level function bar
new binary (where the profile is applied to): bar is inlined into foo.

Right now, BOLT does 1:1 matching between profile functions and binary functions based on the name. #100446 will extend this to N:M where multiple profiles can be matched to one binary function (as in the example above where binary function foo would use profiles for foo and bar), and one profile can be matched to multiple binary functions (eg if bar was inlined into multiple functions).

aaupov committed:
[𝘀𝗽𝗿] changes introduced through rebase

Created using spr 1.3.4

[skip ci]

aaupov committed:
address comments

Created using spr 1.3.4

aaupov committed:
[𝘀𝗽𝗿] changes introduced through rebase

Created using spr 1.3.4

[skip ci]

aaupov committed:
rebase

Created using spr 1.3.4

wlei-llvm commented:
> > > > Ping @wlei-llvm
> > > 
> > > 
> > > Sorry for the delay. The new version addressed my last comment (with just minor nits). However, I didn't fully follow the new features related to `ProbeMatchSpecs` stuffs. Could you add more descriptions to the diff summary? Or if it’s not a lot of work, could we split it into two patches? We could commit the first part, and I will review the second part separately.
> > 
> > 
> > NVM, I think now I get what `ProbeMatchSpecs` does, it's a vector because a function can have multiple sections(function split)
> 
> Thank you for reviewing and sorry for the delay from my end, was busy with profile quality work.
> 
> ProbeMatchSpecs is a mechanism to match probes belonging to another binary function. I'm going to utilize it in probe-based function matching (#100446). For example:
> 
> source function:
> 
> ```
> void foo() {
>   bar();
> }
> ```
> 
> profiled binary: bar is not inlined => have top-level function bar new binary (where the profile is applied to): bar is inlined into foo.
> 
> Right now, BOLT does 1:1 matching between profile functions and binary functions based on the name. #100446 will extend this to N:M where multiple profiles can be matched to one binary function (as in the example above where binary function foo would use profiles for foo and bar), and one profile can be matched to multiple binary functions (eg if bar was inlined into multiple functions).
> 
> In this diff, ProbeMatchSpecs would only have one BinaryFunctionProfile (existing name-based matching).

Thanks for the explanation!  I was confused of the use of `ProbeMatchSpecs`, it would be great to add more descriptions in the diff summary or somewhere in the comments, or in the another patch when it's used(IMO if we add a feature, but we doesn't use it in the patch, it should be in the future patch when it's used)

wlei-llvm reviewed (approved):
LGTM, thanks for addressing the comments!

aaupov commented:
> > > > > Ping @wlei-llvm
> > > > 
> > > > 
> > > > Sorry for the delay. The new version addressed my last comment (with just minor nits). However, I didn't fully follow the new features related to `ProbeMatchSpecs` stuffs. Could you add more descriptions to the diff summary? Or if it’s not a lot of work, could we split it into two patches? We could commit the first part, and I will review the second part separately.
> > > 
> > > 
> > > NVM, I think now I get what `ProbeMatchSpecs` does, it's a vector because a function can have multiple sections(function split)
> > 
> > 
> > Thank you for reviewing and sorry for the delay from my end, was busy with profile quality work.
> > ProbeMatchSpecs is a mechanism to match probes belonging to another binary function. I'm going to utilize it in probe-based function matching (#100446). For example:
> > source function:
> > ```
> > void foo() {
> >   bar();
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > profiled binary: bar is not inlined => have top-level function bar new binary (where the profile is applied to): bar is inlined into foo.
> > Right now, BOLT does 1:1 matching between profile functions and binary functions based on the name. #100446 will extend this to N:M where multiple profiles can be matched to one binary function (as in the example above where binary function foo would use profiles for foo and bar), and one profile can be matched to multiple binary functions (eg if bar was inlined into multiple functions).
> > In this diff, ProbeMatchSpecs would only have one BinaryFunctionProfile (existing name-based matching).
> 
> Thanks for the explanation! I was confused of the use of `ProbeMatchSpecs`, it would be great to add more descriptions in the diff summary or somewhere in the comments, or in the another patch when it's used(IMO if we add a feature, but we doesn't use it in the patch, it should be in the future patch when it's used)

Thank you. Added description of ProbeMatchSpecs to the summary.
I decided to introduce it in this diff because there's tight coupling between probe-based block matching and function matching. This way, probe-based function matching would not need to change how block matching works.

wlei-llvm commented:
> > > > > > Ping @wlei-llvm
> > > > > 
> > > > > 
> > > > > Sorry for the delay. The new version addressed my last comment (with just minor nits). However, I didn't fully follow the new features related to `ProbeMatchSpecs` stuffs. Could you add more descriptions to the diff summary? Or if it’s not a lot of work, could we split it into two patches? We could commit the first part, and I will review the second part separately.
> > > > 
> > > > 
> > > > NVM, I think now I get what `ProbeMatchSpecs` does, it's a vector because a function can have multiple sections(function split)
> > > 
> > > 
> > > Thank you for reviewing and sorry for the delay from my end, was busy with profile quality work.
> > > ProbeMatchSpecs is a mechanism to match probes belonging to another binary function. I'm going to utilize it in probe-based function matching (#100446). For example:
> > > source function:
> > > ```
> > > void foo() {
> > >   bar();
> > > }
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > profiled binary: bar is not inlined => have top-level function bar new binary (where the profile is applied to): bar is inlined into foo.
> > > Right now, BOLT does 1:1 matching between profile functions and binary functions based on the name. #100446 will extend this to N:M where multiple profiles can be matched to one binary function (as in the example above where binary function foo would use profiles for foo and bar), and one profile can be matched to multiple binary functions (eg if bar was inlined into multiple functions).
> > > In this diff, ProbeMatchSpecs would only have one BinaryFunctionProfile (existing name-based matching).
> > 
> > 
> > Thanks for the explanation! I was confused of the use of `ProbeMatchSpecs`, it would be great to add more descriptions in the diff summary or somewhere in the comments, or in the another patch when it's used(IMO if we add a feature, but we doesn't use it in the patch, it should be in the future patch when it's used)
> 
> Thank you. Added description of ProbeMatchSpecs to the summary. I decided to introduce it in this diff because there's tight coupling between probe-based block matching and function matching. This way, probe-based function matching would not need to change how block matching works.

Makes sense, thanks for clarifying!

kazutakahirata commented:
I've fixed a couple of warnings from this PR with https://github.com/llvm/llvm-project/commit/06e08696248ac01754c87c22cc8a4b797ef46430.  Thanks!
